Compile xcpLite under Ubuntu

Hello,

I tried to compile the CMake project, but got several errors. 
Could you please fix this? 

I will attach later on the errors.

Thanks

Hi, sorry for my late reply.

Following errors occur when compiling the cmake project under Ubuntu /WSL.

```
~/xcpTl.c:421:46: error: invalid conversion from ‘char*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
  421 |         n = socketRecv(gXcpTl.MulticastSock, (char*) buffer, (uint16_t)sizeof(buffer));
```

```
/usr/bin/ld: CMakeFiles/xcpLite.dir/xcpLite.c.o: in function `XcpCommand':
~/xcpLite.c:1116: undefined reference to `XcpTlSetClusterId'
/usr/bin/ld: CMakeFiles/xcpLite.dir/xcpLite.c.o: in function `XcpInit':
~/xcpLite.c:1349: undefined reference to `XcpTlSetClusterId'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/xcpLite.dir/build.make:219: xcpLite] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/xcpLite.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

> Hi, sorry for my late reply.
> 
> Following errors occur when compiling the cmake project under Ubuntu /WSL.
> 
> ```
> ~/xcpTl.c:421:46: error: invalid conversion from ‘char*’ to ‘uint8_t*’ {aka ‘unsigned char*’} [-fpermissive]
>   421 |         n = socketRecv(gXcpTl.MulticastSock, (char*) buffer, (uint16_t)sizeof(buffer));
> ```
> 
> ```
> /usr/bin/ld: CMakeFiles/xcpLite.dir/xcpLite.c.o: in function `XcpCommand':
> ~/xcpLite.c:1116: undefined reference to `XcpTlSetClusterId'
> /usr/bin/ld: CMakeFiles/xcpLite.dir/xcpLite.c.o: in function `XcpInit':
> ~/xcpLite.c:1349: undefined reference to `XcpTlSetClusterId'
> collect2: error: ld returned 1 exit status
> make[2]: *** [CMakeFiles/xcpLite.dir/build.make:219: xcpLite] Error 1
> make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/xcpLite.dir/all] Error 2
> make: *** [Makefile:84: all] Error 2
> ```
> 
> Thank you!

You can use the patch from the pull-request.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,15 +1,10 @@
 cmake_minimum_required(VERSION 3.1.0)
 project(xcpLite)
 
-set(CMAKE_C_COMPILER "g++")
-set(CMAKE_CXX_COMPILER "g++")
-
 set(SOURCES main.cpp ecupp.cpp ecu.c xcpAppl.c xcpLite.c xcpTl.c A2L.c xcpSlave.c clock.c platform.c)
-set(CMAKE_BUILD_TYPE Debug)
 
 add_executable(xcpLite ${SOURCES})
 
 set(THREADS_PREFER_PTHREAD_FLAG ON)
 find_package(Threads REQUIRED)
 target_link_libraries(xcpLite PRIVATE Threads::Threads)
-

--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@ Copyright 2021 Vector Informatik GmbH
 
 Lightweight implementation of the ASAM XCP Protocol Layer V1.4 (1000 lines of code in XCPlite.c).
 
-Supports Linux 32/64 Bit and Windows 32/64 Bit 
+Supports Linux 32/64 Bit and Windows 32/64 Bit
 Posix and Windows Sockets or Vector VN5xxx automotive Ethernet devices
 100-Base-T1 or 1000-Base-T1 BroadrReach, XL-API V3 network based access
 
@@ -17,8 +17,8 @@ C and C++ target support.
 
 Achieves 40 MByte/s throughput on a Raspberry Pi 4 (jumbo frames enabled) with 3% cpu time in event copy routine
 
-Quick start with no A2L (ASAP2 ECU description) required. 
-An A2L with reduced featureset is generated through code instrumentation during runtime on target system 
+Quick start with no A2L (ASAP2 ECU description) required.
+An A2L with reduced featureset is generated through code instrumentation during runtime on target system
 and automatically uploaded by XCP).
 
 C and C++ measurement demo variables and code example in ecu.c and ecupp.cpp.
@@ -40,7 +40,7 @@ Example:
 
 ```
   channel1 = 0;
-  
+
   A2lCreateEvent("ECU"); // Create a new event with name "ECU""
   A2lSetEvent("ECU"); // Set event "ECU" to be associated to following measurement definitions
   A2lCreatePhysMeasurement(channel1, 2.0, 1.0, "Volt", "Demo floating point signal"); // Create a new measurement signal "channel1" with linear conversion rule (factor,offset) and unit "Volt"
@@ -54,7 +54,7 @@ Example:
   XcpEvent(1); // Trigger event here, timestamp and copy measurement data
 ```
 
-Demo visual Studio solution and CANape project included for Raspberry Pi 4 and Windows 32/64. 
+Demo visual Studio solution and CANape project included for Raspberry Pi 4 and Windows 32/64.
 
 ![CANape](Screenshot.png)
 
@@ -67,10 +67,10 @@ All settings and parameters for the XCP protocol and transport layer are located
 Compile options for the XCPlite demo are located in main_cfg.h:
 
 ## Notes:
-- If A2L generation and upload is disabled, use CANape address update with Linker Map Type ELF extended for a.out format or PDB for .exe 
-- The A2L generator creates a unique file name for the A2L, for convinience use name detection (GET_ID 1) 
+- If A2L generation and upload is disabled, use CANape address update with Linker Map Type ELF extended for a.out format or PDB for .exe
+- The A2L generator creates a unique file name for the A2L, for convinience use name detection (GET_ID 1)
 - Linux Compile with -O2, Link with -pthread
-- 64 bit version needs all objects within one 4 GByte data segment  
+- 64 bit version needs all objects within one 4 GByte data segment
 
 ## Version History
 Version 4:
@@ -78,21 +78,48 @@ Version 4:
 - All dependencies to UDP socket library in platform.h/.c
 - Support for Vector XL-API removed
 
+## Build
 
+### Linux x86_64
 
+$ sudo apt-get install cmake g++ clang cmake ninja-build
 
+#### Release Build
+$ cd XcpLite
+$ mkdir build_release
+$ cd build_release
+$ cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
+$ ninja
 
+#### Debug Build
+$ cd XcpLite
+$ mkdir build_debug
+$ cd build_debug
+$ cmake -GNinja ..
+$ ninja
 
+### Windows x86_64
 
+Use the Visual Studio projects or CMake.
 
+For the CMake setup, prepare your command line environment.
+Set compiler to Microsoft x64 cl.exe and make sure the system finds cmake.exe and ninja.exe.
+You can also use the Windows clang compiler.
 
+> call "C:\Program Files (x86)\Microsoft Visual Studio 15.0\VC\Auxiliary\Build\vcvars64.bat"
+> set PATH=C:\Tools\ninja;%PATH%
+> set PATH=C:\Tools\cmake_3.17.2.0\bin;%PATH%
 
+#### Release Build
+> cd XcpLite
+> mkdir build_release
+> cd build_release
+> cmake.exe -GNinja -DCMAKE_BUILD_TYPE=Release ..
+> ninja.exe
 
-
-
-
-
-
-
-
-
+#### Debug Build
+> cd XcpLite
+> mkdir build_debug
+> cd build_debug
+> cmake.exe -GNinja ..
+> ninja.exe

--- a/README.md
+++ b/README.md
@@ -85,19 +85,22 @@ Version 4:
 $ sudo apt-get install cmake g++ clang cmake ninja-build
 
 #### Release Build
+```
 $ cd XcpLite
 $ mkdir build_release
 $ cd build_release
 $ cmake -GNinja -DCMAKE_BUILD_TYPE=Release ..
 $ ninja
+```
 
 #### Debug Build
+```
 $ cd XcpLite
 $ mkdir build_debug
 $ cd build_debug
 $ cmake -GNinja ..
 $ ninja
-
+```
 ### Windows x86_64
 
 Use the Visual Studio projects or CMake.
@@ -105,21 +108,24 @@ Use the Visual Studio projects or CMake.
 For the CMake setup, prepare your command line environment.
 Set compiler to Microsoft x64 cl.exe and make sure the system finds cmake.exe and ninja.exe.
 You can also use the Windows clang compiler.
-
+```
 > call "C:\Program Files (x86)\Microsoft Visual Studio 15.0\VC\Auxiliary\Build\vcvars64.bat"
 > set PATH=C:\Tools\ninja;%PATH%
 > set PATH=C:\Tools\cmake_3.17.2.0\bin;%PATH%
-
+```
 #### Release Build
+```
 > cd XcpLite
 > mkdir build_release
 > cd build_release
 > cmake.exe -GNinja -DCMAKE_BUILD_TYPE=Release ..
 > ninja.exe
-
+```
 #### Debug Build
+```
 > cd XcpLite
 > mkdir build_debug
 > cd build_debug
 > cmake.exe -GNinja ..
 > ninja.exe
+```

--- a/clock.c
+++ b/clock.c
@@ -30,7 +30,7 @@ uint64_t clockGetLast64() {
 
 #ifdef _LINUX // Linux
 
-/* 
+/*
 Linux clock type
   CLOCK_REALTIME  This clock is affected by incremental adjustments performed by NTP
   CLOCK_TAI       This clock does not experience discontinuities and backwards jumps caused by NTP inserting leap seconds as CLOCK_REALTIME does.
@@ -59,20 +59,20 @@ char* clockGetString(char* s, unsigned int cs, uint64_t c) {
 
 
 int clockInit()
-{    
+{
 
     printf("\nInit clock\n  (");
-#ifdef CLOCK_USE_UTC_TIME_US 
+#ifdef CLOCK_USE_UTC_TIME_US
     #error "CLOCK_USE_UTC_TIME_US is deprecated!");
 #endif
-#ifdef CLOCK_USE_UTC_TIME_NS 
+#ifdef CLOCK_USE_UTC_TIME_NS
     printf("CLOCK_USE_UTC_TIME_NS,");
 #endif
-#ifdef CLOCK_USE_APP_TIME_US 
+#ifdef CLOCK_USE_APP_TIME_US
     printf("CLOCK_USE_APP_TIME_US,");
 #endif
 
-#if CLOCK_TYPE == CLOCK_TAI 
+#if CLOCK_TYPE == CLOCK_TAI
     printf("CLOCK_TYPE_TAI,");
 #endif
 #if CLOCK_TYPE == CLOCK_REALTIME
@@ -102,8 +102,8 @@ int clockInit()
         gettimeofday(&ptm, NULL);
         clock_gettime(CLOCK_TAI, &gts_TAI);
         clock_gettime(CLOCK_REALTIME, &gts_REALTIME);
-        printf("  CLOCK_TAI=%us CLOCK_REALTIME=%us time=%u timeofday=%u\n", gts_TAI.tv_sec, gts_REALTIME.tv_sec, now, ptm.tv_sec);
-        // Check 
+        printf("  CLOCK_TAI=%lu CLOCK_REALTIME=%lu time=%lu timeofday=%lu\n", gts_TAI.tv_sec, gts_REALTIME.tv_sec, now, ptm.tv_sec);
+        // Check
         t1 = clockGet64();
         sleepNs(100000);
         t2 = clockGet64();
@@ -119,7 +119,7 @@ int clockInit()
 // Free running clock with 1us tick
 uint32_t clockGet32() {
 
-    struct timespec ts; 
+    struct timespec ts;
     clock_gettime(CLOCK_TYPE, &ts);
 #ifdef CLOCK_USE_UTC_TIME_NS // ns since 1.1.1970
     gClock64 = (((uint64_t)(ts.tv_sec) * 1000000000ULL) + (uint64_t)(ts.tv_nsec)); // ns
@@ -127,7 +127,7 @@ uint32_t clockGet32() {
     gClock64 = (((uint64_t)(ts.tv_sec-gts0.tv_sec) * 1000000ULL) + (uint64_t)(ts.tv_nsec / 1000)); // us
 #endif
     gClock32 = (uint32_t)gClock64;
-    return gClock32;  
+    return gClock32;
 }
 
 uint64_t clockGet64() {
@@ -153,7 +153,7 @@ void sleepMs(uint32_t ms) {
 
 #else // Windows
 
-// Performance counter to clock conversion 
+// Performance counter to clock conversion
 static uint64_t sFactor = 0; // ticks per us
 static uint8_t sDivide = 0; // divide or multiply
 static uint64_t sOffset = 0; // offset
@@ -179,7 +179,7 @@ char *clockGetString(char* s, unsigned int cs, uint64_t c) {
 int clockInit() {
 
     printf("\nInit clock\n");
-#ifdef CLOCK_USE_UTC_TIME_NS 
+#ifdef CLOCK_USE_UTC_TIME_NS
     printf("  CLOCK_USE_UTC_TIME_NS\n");
 #endif
 
@@ -230,11 +230,11 @@ int clockInit() {
     tp = (((int64_t)tC.u.HighPart) << 32) | (int64_t)tC.u.LowPart;
 #ifndef CLOCK_USE_UTC_TIME_NS
     // Reset clock now
-    sOffset = tp; 
+    sOffset = tp;
 #else
     // set  offset from local clock UTC value t
     // this is inaccurate up to 1 s, but irrelevant because system clock UTC offset is also not accurate
-    sOffset = (uint64_t)t * CLOCK_TICKS_PER_S  + t_ms * CLOCK_TICKS_PER_MS - tp * sFactor; 
+    sOffset = (uint64_t)t * CLOCK_TICKS_PER_S  + t_ms * CLOCK_TICKS_PER_MS - tp * sFactor;
 #endif
 
     clockGet64();
@@ -260,7 +260,7 @@ int clockInit() {
 
 // Clock 64 Bit (UTC or ARB)
 uint64_t clockGet64() {
-   
+
     LARGE_INTEGER tp;
     uint64_t t;
 
@@ -315,5 +315,3 @@ void sleepMs(unsigned int ms) {
 
 
 #endif // Windows
-
-

--- a/xcpAppl.c
+++ b/xcpAppl.c
@@ -3,7 +3,7 @@
 |   xcpAppl.c
 |
 | Description:
-|   Externals for xcpLite 
+|   Externals for xcpLite
 |   Platform and implementation specific functions
 |   All other callbacks/dependencies are implemented as macros in xcpAppl.h
 |
@@ -43,7 +43,7 @@ uint8_t ApplXcpGetClockInfo(T_CLOCK_INFO_SLAVE* s, T_CLOCK_INFO_GRANDMASTER* m) 
 
     uint8_t uuid[8] = APP_DEFAULT_CLOCK_UUID;
 
-    
+
         memcpy(s->UUID, uuid, 8);
         memcpy(m->UUID, uuid, 8);
 #ifdef CLOCK_USE_UTC_TIME_NS
@@ -55,7 +55,7 @@ uint8_t ApplXcpGetClockInfo(T_CLOCK_INFO_SLAVE* s, T_CLOCK_INFO_GRANDMASTER* m) 
         m->stratumLevel = XCP_STRATUM_LEVEL_UNSYNC;
         m->epochOfGrandmaster = XCP_EPOCH_ARB;
 #endif
-    
+
 
 
     if (gDebugLevel >= 1) {
@@ -141,7 +141,7 @@ static int dump_phdr(struct dl_phdr_info* pinfo, size_t size, void* data)
 }
 
 uint8_t* ApplXcpGetBaseAddr() {
-  
+
   if (!baseAddrValid) {
     dl_iterate_phdr(dump_phdr, NULL);
     assert(baseAddr!=NULL);
@@ -169,7 +169,7 @@ uint8_t* ApplXcpGetPointer(uint8_t addr_ext, uint32_t addr)
 
 #ifdef _LINUX32
 
-uint8_t* ApplXcpGetBaseAddr() 
+uint8_t* ApplXcpGetBaseAddr()
 {
 
     return ((uint8_t*)0);
@@ -234,7 +234,7 @@ uint8_t ApplXcpGetSlaveId(char** p, uint32_t* n) {
 
 static char gA2LFilename[100]; // Name without extension
 
-// A2L base name for GET_ID 
+// A2L base name for GET_ID
 static char gA2LPathname[MAX_PATH + 100 + 4]; // Full path + name +extension
 
 
@@ -302,7 +302,7 @@ uint8_t ApplXcpReadFile(uint8_t type, uint8_t** p, uint32_t* n) {
             CloseHandle(hFile);
         }
         MultiByteToWideChar(0, 0, filename, (int)strlen(filename), wcfilename, (int)strlen(filename));
-        hFile = CreateFile(wcfilename, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+        hFile = CreateFileW((wchar_t*)wcfilename, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
         if (hFile == INVALID_HANDLE_VALUE) {
             printf("file %s not found!\n", filename);
             return 0;
@@ -316,15 +316,11 @@ uint8_t ApplXcpReadFile(uint8_t type, uint8_t** p, uint32_t* n) {
 #ifdef XCP_ENABLE_TESTMODE
             if (gDebugLevel >= 1) printf("  file %s ready for upload, size=%u, mta=%p\n\n", filename, gXcpFileLength, gXcpFile);
 #endif
-        
-    
+
+
     *n = gXcpFileLength;
     *p = gXcpFile;
     return 1;
 }
 
 #endif
-
-
-
-

--- a/xcpTl.c
+++ b/xcpTl.c
@@ -35,7 +35,7 @@ static struct {
     uint16_t MasterPort;
     int MasterAddrValid;
 
-    // Transmit queue 
+    // Transmit queue
     tXcpDtoBuffer dto_queue[XCPTL_DTO_QUEUE_SIZE];
     unsigned int dto_queue_rp; // rp = read index
     unsigned int dto_queue_len; // rp+len = write index (the next free entry), len=0 ist empty, len=XCPTL_DTO_QUEUE_SIZE is full
@@ -47,13 +47,13 @@ static struct {
     uint16_t CrmCtr; // next CRM command response message packet counter
 
     // DTO data transfer object counter (DAQ,STIM)
-    uint16_t DtoCtr; // next DAQ DTO data transmit message packet counter 
+    uint16_t DtoCtr; // next DAQ DTO data transmit message packet counter
 
     // Multicast
 #ifdef XCPTL_ENABLE_MULTICAST
     tXcpThread MulticastThreadHandle;
     SOCKET MulticastSock;
-#endif 
+#endif
 
     MUTEX Mutex_Queue;
     MUTEX Mutex_Send;
@@ -78,7 +78,7 @@ static int handleXcpMulticast(int n, tXcpCtoMessage* p);
 static int sendDatagram(const uint8_t *data, uint16_t size ) {
 
     int r;
-        
+
 #ifdef XCP_ENABLE_TESTMODE
     if (ApplXcpGetDebugLevel() >= 3) {
         printf("TX: size=%u ",size);
@@ -159,7 +159,7 @@ int XcpTlHandleTransmitQueue( void ) {
         mutexLock(&gXcpTl.Mutex_Queue);
         if (gXcpTl.dto_queue_len > 1) {
             b = &gXcpTl.dto_queue[gXcpTl.dto_queue_rp];
-            if (b->xcp_uncommited > 0) b = NULL; 
+            if (b->xcp_uncommited > 0) b = NULL;
         }
         else {
             b = NULL;
@@ -194,10 +194,10 @@ void XcpTlFlushTransmitQueue() {
 #endif
 
     // Complete the current buffer if non empty
-    mutexLock(&gXcpTl.Mutex_Queue);               
+    mutexLock(&gXcpTl.Mutex_Queue);
     if (gXcpTl.dto_buffer_ptr!=NULL && gXcpTl.dto_buffer_ptr->xcp_size>0) getDtoBuffer();
     mutexUnlock(&gXcpTl.Mutex_Queue);
-        
+
     XcpTlHandleTransmitQueue();
 }
 
@@ -220,7 +220,7 @@ uint8_t *XcpTlGetDtoBuffer(void **par, uint16_t size) {
 #endif
 
     mutexLock(&gXcpTl.Mutex_Queue);
-        
+
     // Get another message buffer from queue, when active buffer ist full, overrun or after time condition
     if (gXcpTl.dto_buffer_ptr==NULL || (uint16_t)(gXcpTl.dto_buffer_ptr->xcp_size + size + XCPTL_TRANSPORT_LAYER_HEADER_SIZE) > gXcpTl.SlaveMTU /*XCPTL_SOCKET_MTU_SIZE*/) {
         getDtoBuffer();
@@ -235,7 +235,7 @@ uint8_t *XcpTlGetDtoBuffer(void **par, uint16_t size) {
         gXcpTl.dto_buffer_ptr->xcp_size = (uint16_t)(gXcpTl.dto_buffer_ptr->xcp_size + size + XCPTL_TRANSPORT_LAYER_HEADER_SIZE);
         *((tXcpDtoBuffer**)par) = gXcpTl.dto_buffer_ptr;
         gXcpTl.dto_buffer_ptr->xcp_uncommited++;
-        
+
     }
     else {
         p = NULL; // Overflow
@@ -257,7 +257,7 @@ void XcpTlCommitDtoBuffer(void *par) {
         if (ApplXcpGetDebugLevel() >= 4) {
             printf("CommitPacketBuffer() c=%u,s=%u\n", p->xcp_uncommited, p->xcp_size);
         }
-#endif   
+#endif
 
         mutexLock(&gXcpTl.Mutex_Queue);
         p->xcp_uncommited--;
@@ -284,7 +284,7 @@ int XcpTlSendCrm(const uint8_t *packet, uint16_t size) {
     return sendDatagram((unsigned char*)&p, (uint16_t)(size + XCPTL_TRANSPORT_LAYER_HEADER_SIZE));
 }
 
-    
+
 static int handleXcpCommands(uint16_t n, tXcpCtoMessage * p, uint8_t *srcAddr, uint16_t srcPort) {
 
     int connected;
@@ -304,11 +304,11 @@ static int handleXcpCommands(uint16_t n, tXcpCtoMessage * p, uint8_t *srcAddr, u
 
         /* Connected */
         if (connected) {
-            
-            // Check src addr 
+
+            // Check src addr
             assert(gXcpTl.MasterAddrValid);
 
-            // Check unicast ip address, not allowed to change 
+            // Check unicast ip address, not allowed to change
             if (memcmp(&gXcpTl.MasterAddr, srcAddr, sizeof(gXcpTl.MasterAddr)) != 0) { // Message from different master received
                 printf("WARNING: message from unknown new master %u.%u.%u.%u, disconnecting!\n", srcAddr[0], srcAddr[1], srcAddr[2], srcAddr[3]);
                 XcpDisconnect();
@@ -316,7 +316,7 @@ static int handleXcpCommands(uint16_t n, tXcpCtoMessage * p, uint8_t *srcAddr, u
                 return 1; // Disconnect
             }
 
-            // Check unicast master udp port, not allowed to change 
+            // Check unicast master udp port, not allowed to change
             if (gXcpTl.MasterPort != srcPort) {
                 printf("WARNING: master port changed from %u to %u, disconnecting!\n", gXcpTl.MasterPort, srcPort);
                 XcpDisconnect();
@@ -331,7 +331,7 @@ static int handleXcpCommands(uint16_t n, tXcpCtoMessage * p, uint8_t *srcAddr, u
         else {
             /* Check for CONNECT command ? */
             const tXcpCto* pCmd = (const tXcpCto*)&p->data[0];
-            if (p->dlc == 2 && CRO_CMD == CC_CONNECT) { 
+            if (p->dlc == 2 && CRO_CMD == CC_CONNECT) {
                 memcpy(gXcpTl.MasterAddr,srcAddr,sizeof(gXcpTl.MasterAddr)); // Save master address, so XcpCommand can send the CONNECT response
                 gXcpTl.MasterPort = srcPort;
                 gXcpTl.MasterAddrValid = 1;
@@ -343,7 +343,7 @@ static int handleXcpCommands(uint16_t n, tXcpCtoMessage * p, uint8_t *srcAddr, u
             }
 #endif
         }
-       
+
         // Actions after successfull connect
         if (!connected) {
             if (XcpIsConnected()) { // Is in connected state
@@ -352,12 +352,12 @@ static int handleXcpCommands(uint16_t n, tXcpCtoMessage * p, uint8_t *srcAddr, u
 #endif
 
                 // Inititialize the DAQ message queue
-                XcpTlInitTransmitQueue(); 
+                XcpTlInitTransmitQueue();
 
-            }  
+            }
             else { // Is not in connected state
                 gXcpTl.MasterAddrValid = 0; // Any client can connect
-            } 
+            }
         } // !connected before
 
     }
@@ -387,7 +387,7 @@ int XcpTlHandleCommands() {
             return 1; // Ok, no command pending
         }
         printf("ERROR %u: recvfrom failed (result=%d)!\n", socketGetLastError(), n);
-        return 0; // Error           
+        return 0; // Error
     }
     else {
         return handleXcpCommands((uint16_t)n, (tXcpCtoMessage*)buffer, srcAddr, srcPort);
@@ -418,7 +418,7 @@ extern void* XcpTlMulticastThread(void* par)
     uint8_t buffer[256];
     int16_t n;
     for (;;) {
-        n = socketRecv(gXcpTl.MulticastSock, (char*)&buffer, (uint16_t)sizeof(buffer));
+        n = socketRecv(gXcpTl.MulticastSock, (uint8_t*)&buffer, (uint16_t)sizeof(buffer));
         if (n < 0) break; // Terminate on error (socket close is used to terminate thread)
         handleXcpMulticast(n, (tXcpCtoMessage*)buffer);
     }
@@ -499,7 +499,7 @@ void XcpTlWaitForTransmitData(unsigned int timeout_us) {
     if (gXcpTl.dto_queue_len <= 1) {
         sleepNs(timeout_us*1000);
     }
-    return; 
+    return;
 }
 
 
@@ -536,6 +536,3 @@ int XcpTlGetSlaveMAC(uint8_t *mac) {
 
     return socketGetLocalAddr(mac,NULL);
 }
-
-
-

